ceph/mds.go

```diff
@@ -183,3 +217,204 @@ func (m *MDSCollector) Collect(ch chan<- prometheus.Metric, version *Version) {
 		}
 	}
 }
+
+type healthDetailCheck struct {
+	Status string `json:"status"`
+	Checks map[string]struct {
+		Severity string `json:"severity"`
+		Summary  struct {
+			Message string `json:"message"`
+			Count   int    `json:"count"`
+		} `json:"summary"`
+		Detail []struct {
+			Message string `json:"message"`
+		} `json:"detail"`
+		Muted bool `json:"muted"`
+	} `json:"checks"`
+}
+
+type mdsStatus struct {
+	ClusterFsid        string  `json:"cluster_fsid"`
+	Whoami             int     `json:"whoami"`
+	ID                 int64   `json:"id"`
+	WantState          string  `json:"want_state"`
+	State              string  `json:"state"`
+	FsName             string  `json:"fs_name"`
+	RankUptime         float64 `json:"rank_uptime"`
+	MdsmapEpoch        int     `json:"mdsmap_epoch"`
+	OsdmapEpoch        int     `json:"osdmap_epoch"`
+	OsdmapEpochBarrier int     `json:"osdmap_epoch_barrier"`
+	Uptime             float64 `json:"uptime"`
+}
+
+type mdsSlowOp struct {
+	Ops []struct {
+		// Custom fields for easy parsing by caller.
+		MDSName, CephFSOpType string
+
+		// CephFS fields.
+		Description string  `json:"description"`
+		InitiatedAt string  `json:"initiated_at"`
+		Age         float64 `json:"age"`
+		Duration    float64 `json:"duration"`
+		TypeData    struct {
+			FlagPoint  string `json:"flag_point"`
+			Reqid      string `json:"reqid"`
+			OpType     string `json:"op_type"`
+			ClientInfo struct {
+				Client string `json:"client"`
+				Tid    int    `json:"tid"`
+			} `json:"client_info"`
+			Events []struct {
+				Time  string `json:"time"`
+				Event string `json:"event"`
+			} `json:"events"`
+		} `json:"type_data,omitempty"`
+	} `json:"ops"`
+	ComplaintTime int `json:"complaint_time"`
+	NumBlockedOps int `json:"num_blocked_ops"`
+}
+
+func (m *MDSCollector) collectMDSSlowOps() {
+	ctx, cancel := context.WithTimeout(context.Background(), 1*time.Minute)
```

Comment:
does the timeout correctly kill any ceph processes that get stuck?

Reply:
It calls `os.Process.Kill()` so I sure hope it does unless ceph commands somehow have a power to persist across that. 😄